network/p2p/topic_validator_test.go

```diff
@@ -109,31 +107,349 @@ func TestTopicValidator(t *testing.T) {
+	header := unittest.BlockHeaderFixture()
+	data1 := getMsgFixtureBz(t, &messages.BlockProposal{Header: &header})
+
+	// sn2 publishes the block proposal, sn1 and an1 should receive the message because
+	// SN nodes are authorized to send block proposals
+	err = sn2.Publish(timedCtx, topic, data1)
+	require.NoError(t, err)
+
+	// sn1 gets the message
+	checkReceive(timedCtx, t, data1, sub1, nil, true)
+
+	// sn2 also gets the message (as part of the libp2p loopback of published topic messages)
+	checkReceive(timedCtx, t, data1, sub2, nil, true)
+
+	// an1 also gets the message
+	checkReceive(timedCtx, t, data1, sub3, nil, true)
+
+	timedCtx, cancel2s := context.WithTimeout(context.Background(), 2*time.Second)
+	defer cancel2s()
+	header = unittest.BlockHeaderFixture()
+	data2 := getMsgFixtureBz(t, &messages.BlockProposal{Header: &header})
+
+	// the access node now publishes the block proposal message, AN are not authorized to publish block proposals
+	// the message should be rejected by the topic validator on sn1
+	err = an1.Publish(timedCtx, topic, data2)
+	require.NoError(t, err)
+
+	// an1 receives its own message
+	checkReceive(timedCtx, t, data2, sub3, nil, true)
+
+	var wg sync.WaitGroup
+
+	// sn1 does NOT receive the message due to the topic validator
+	timedCtx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	checkReceive(timedCtx, t, nil, sub1, &wg, false)
+
+	// sn2 also does not receive the message via gossip from the sn1 (event after the 1 second hearbeat)
+	timedCtx, cancel = context.WithTimeout(context.Background(), 2*time.Second)
+	defer cancel()
+	checkReceive(timedCtx, t, nil, sub2, &wg, false)
+
+	unittest.RequireReturnsBefore(t, wg.Wait, 5*time.Second, "could not receive message on time")
+
+	// expecting 1 warn calls for each rejected message from unauthorized node
+	require.Equalf(t, 1, hookCalls, "expected 1 warning to be logged")
+}
+
+// TestAuthorizedSenderValidator_Authorized tests that the authorized sender validator rejects messages from nodes that are ejected
+func TestAuthorizedSenderValidator_Ejected(t *testing.T) {
+	sporkId := unittest.IdentifierFixture()
+	identity1, privateKey1 := unittest.IdentityWithNetworkingKeyFixture(unittest.WithRole(flow.RoleConsensus))
+	sn1 := createNode(t, identity1.NodeID, privateKey1, sporkId)
+
+	identity2, privateKey2 := unittest.IdentityWithNetworkingKeyFixture(unittest.WithRole(flow.RoleConsensus))
+	sn2 := createNode(t, identity2.NodeID, privateKey2, sporkId)
+
+	identity3, privateKey3 := unittest.IdentityWithNetworkingKeyFixture(unittest.WithRole(flow.RoleAccess))
+	an1 := createNode(t, identity3.NodeID, privateKey3, sporkId)
+
+	topic := engine.TopicFromChannel(engine.SyncCommittee, sporkId)
+
+	ids := flow.IdentityList{identity1, identity2, identity3}
+	translator, err := NewFixedTableIdentityTranslator(ids)
+	require.NoError(t, err)
+
+	// setup hooked logger
+	hookCalls := 0
```

Comment:
I think we should use a synchronized variable here as well (same suggestion as previous test)

Reply:
https://github.com/onflow/flow-go/pull/2317/commits/079042c5ef705456cae43752bd375f29afdec63c